roles/kubernetes/preinstall/tasks/0020-verify-settings.yml

```diff
@@ -108,6 +108,7 @@
     msg: "'{{ ansible_all_ipv4_addresses }}' do not contain '{{ ip }}'"
   when:
     - not ignore_assert_errors
+    - ansible_all_ipv4_address is defined
```

Comment:
```suggestion
    - ansible_all_ipv4_addresses is defined
```
Is that correct?

Reply:
Nope, doesn't work. With `ansible_all_ipv4_address` on the other hand it runs and fails only at #8881.
So, I guess that this fix will do its job.

Reply:
There are not any places to define/use `ansible_all_ipv4_address` in Kubespray.
I am not sure why current pull request fixes the issue.

Reply:
I'm guessing the current PR actually just masks an issue because of the typo. The `ansible_all_ipv4_address` is always defined but is an empty list in ipv6 only hosts so the check should be something like `ansible_all_ipv4_address` which evaluates to `False` for an empty list.

Reply:
> I'm guessing the current PR actually just masks an issue because of the typo.

oh, that is a nice point..

Reply:
Well, it works ^^
But it could have a side effect for IPv4 users, I guess?

Reply:
> Well, it works ^^
> But it could have a side effect for IPv4 users, I guess?

Could you elaborate? What kind of side effect do you see?

Reply:
`ansible_facts` gathers variables including `ansible_all_ipv4_addresses`, not `ansible_all_ipv4_address`.
Then we are saying `ansible_all_ipv4_address` would be a typo.
I am not sure @Citrullin understands it correctly.

Ref: https://docs.ansible.com/ansible/latest/user_guide/playbooks_vars_facts.html

Reply:
> > Well, it works ^^
> > But it could have a side effect for IPv4 users, I guess?
> 
> Could you elaborate? What kind of side effect do you see?

I don't use legacy IP, so I can't test it. All I care about is making the current IP standard work. I am not too concerned about the internal functionality. All I can tell you is: `ansible_all_ipv4_addresses` is failing, while `ansible_all_ipv4_address` isn't.

Reply:
Sadly, as maintainers we do care about compatibility and there is an implicit expectations that folks who want to have their particular issue fixed help in considering the wider picture before expecting their contributions to be accepted.

As @oomichi said, there is a typo in your code and the reason this change works for you is that the typo hides an issue.

Reply:
> Sadly, as maintainers we do care about compatibility

Well, you apparently haven't tested it with IPv6. Which is the current IP standard. So much to compatibility. As I said. I would test it on legacy IP, but I can't. I literally don't have access to a legacy IP cluster. It's just driving me mad that I have to fix every other project, because no one implements current standards that even exist for over a decade. 

> As @oomichi said, there is a typo in your code and the reason this change works for you is that the typo hides an issue.

Possible. I am not familiar enough with your project to tell you. Can't test it on legacy IP stack. 
I mean you can also ask what is the point of having it in the first place. You could also remove the whole block and it wouldn't change much. Sure, it would fail if there are no IPs given, but people should be able to get that from the error message.

Reply:
@Citrullin IPv6 only is still a rarity and we indeed don't test single-stack IPv6 but for you to validate your proposed change against an IPv4 setup is pretty trivial, you con't need to run a full kubespray to check which of the two variables are correct, just adding a fake vet with `ip link add type veth` and running ansible to see the variables computed by it should give you an idea of how an IPv4 environment looks like.

Reply:
> @Citrullin IPv6 only is still a rarity and we indeed don't test single-stack IPv6 but for you to validate your proposed change against an IPv4 setup is pretty simple, you don't need to run a full kubespray to check which of the two variables are correct, just adding a fake vet with `ip link add type veth` and running ansible to see the variables computed by it should give you an idea of how an IPv4 environment looks like.

Why should I test on a legacy standard that is going to get faced out? IPv4 is optional. Treat it as such. This shit drives me mad. You are professionals, not some kids tinkering around with tech. I expect professional behavior from you. 
https://tools.ietf.org/id/draft-ietf-sunset4-ipv6-ietf-01.html

IPv6 is only a rarity on server side, because developers refuse to move on. On client side it is on it's way to become the majority.
See:
https://www.akamai.com/internet-station/cyber-attacks/state-of-the-internet-report/ipv6-adoption-visualization
https://www.google.de/ipv6/statistics.html#tab=ipv6-adoption

Reply:
@Citrullin Thank you for testing it on your side.
From your test result, `ansible_all_ipv4_addresses is defined` doesn't work.
I guess `ansible_all_ipv4_addresses` is defined and empty(`[]`).
What happens if the following change is applied instead?
```
- ansible_all_ipv4_addresses|length>0
```
That can avoid running the check on environments without IPv4.

Reply:
safest would be
```
- ansible_all_ipv4_addresses | default([]) | length>0
```

Reply:
Thank you for your suggestion @champtar 
@Citrullin Please try @champtar's one, that is better than mine.

Reply:
Sorry for jumping in here, but some corrections needs to be made.

> @Citrullin IPv6 only is still a rarity and we indeed don't test single-stack IPv6 but for you to validate your proposed change against an IPv4 setup is pretty simple, you don't need to run a full kubespray to check which of the two variables are correct, just adding a fake vet with `ip link add type veth` and running ansible to see the variables computed by it should give you an idea of how an IPv4 environment looks like.

What is this?
Sure IPv4 only is still more prominent than IPv6 only, but IPv6 is defined in RFC8200 as current internet standard. If you care about compability as you claim, then you should be at least equally interested in testing IPv6 only.

Calling IPv6 a "rarity" when it is used by over 40% of internet users, well that is foul language and insult. If you expect respectfull language (which we all should) then you need to do the same, and don't call current standards "rarity" !

Reply:
In your 40%, most will be dual stack and not IPv6 only. IPv6 only is pretty rare where you deploy K8S, not sure why you get so upset about it.

Reply:
Guys, calm down.
Nobody is saying that IPV6 support is something minor/useless and not wanted for kubespray.

This PR and all associated issue/PR on IPv6 are important and more than welcome, the CI don't test IPv6 so we probably have a lot of bug/missbehavior when using it.
Sadly current maintainers are not using IPv6 setups, that probably why there is a lot of things missing on that front but again if anybody else can provide support/fix, that's awesome.

I get @Citrullin frustration, as we are maybe one of many projects that don't have extended IPv6 support, but please keep in mind that we are trying as best as we can, on our free time, to maintain every aspect of kubespray up to date.

Reply:
@cristicalin Sorry, this was a little too much. I am just very frustrated that I have to fix every other open source project, because no one gives a damn about IPv6. I am frustrated that no one thinks about the consequences of it. IPv4 addresses are already expensive and they become more expensive over the time. There are tons of companies in emerging markets that will not be able to handle those costs. In the west this seems to us not as such a big deal. 70 USD or even more per address can be huge expenses for other though. I don't like technology that discriminates people, but that is the reality. Our acting or not acting has real consequences to real people. And typically poor people lose. I don't like that. Just to give you some context where the frustration is coming from. Anyway. I am going to look into it and test the different solutions. There are other issues that came up as well.

Reply:
> I guess `ansible_all_ipv4_addresses` is defined and empty(`[]`) on environments where are IPv6 only.

It's not. I don't know what ansible is doing, but there are some IPv4 addresses in there. I haven't defined them though.
Check [this inventory debug dump](http://sprunge.us/5wakwe).